### DISCUSSION TITLE
Suggestion for migrating from pure fastapi project

### DISCUSSION THREAD

**Discussion by zihaom** [zauberzeug/nicegui#4145] [Q&A] [answered]
### Question

Hi, my project is currently based on fastapi+jinja2 templates. I want to utilize the nice agGrid tables generated by nicegui and tried to just generate html (of the table) and embed it in my templates but didnt work. (Got the 'slot stack for this task is empty.' bug ). 

Then I decided to look into the fastapi example, and had some questions. 
1. What does ui.run_with do? Is it like a separated router that hosts all the nicegui stuff (another fastapi app)? Can I pass data (say a pd dataframe) between a fastapi api to the nicegui page?
2. Since nicegui is wrapped around fastapi, is there any point of keeping the old fastapi stuff? like db, schemas, ui.run_with, and just use ui.run()?

**Comment by zihaom** [comment]
In addition, where does the binding of the frontend apis (e.g. show()) happen? during run with? If I want to separate the different pages and import them in frontend.py how would I do it?

**Comment by rodja** [comment, marked as answer]
> Can I pass data (say a pd dataframe) between a fastapi api to the nicegui page?

Yes.

> Since nicegui is wrapped around fastapi, is there any point of keeping the old fastapi stuff? like db, schemas, ui.run_with, and just use ui.run()?

You are right. When migrating from a pure FastAPI implementation, it might me cleaner/easier to just remove your own app instantiation and use `from nicegui import app` instead to register the non-UI routes.

> In addition, where does the binding of the frontend apis (e.g. show()) happen? during run with? If I want to separate the different pages and import them in frontend.py how would I do it?

Have a look at our [modularization example](https://github.com/zauberzeug/nicegui/tree/main/examples/modularization/).

**Reply by zihaom** [reply]
Thanks a lot for your answer!
